src/common/tensor_inspector.h

```diff
@@ -0,0 +1,733 @@
+        for (unsigned int i = 1; i < coords.size(); ++i) {
+          ss << ", " << coords[i];
+        }
+        ss << ")";
+        ret->push_back(coords);
+      }
+    }
+    ss << "]" << std::endl;
+    if (interactive) {
+      std::lock_guard<std::mutex> lock(InspectorManager::get()->mutex_);
+       InspectorManager::get()->check_value_tag_counter_[tag] += 1;
+      while (!InspectorManager::get()->check_value_skip_all_) {
+        std::cout << "----------Value Check----------" << std::endl;
+        if (tag != "") {
+          std::cout << "Tag: " << tag << "  Visit: " <<
+              InspectorManager::get()->check_value_tag_counter_[tag] <<  std::endl;
+        }
+        std::cout << count << " value(s) found." << std::endl;
+        std::cout << "\"p\" to print the coordinates, \"b\" to break, \"s\" to skip all: ";
+        std::string str;
+        std::cin >> str;
+        if (str == "b") {
+          break;
+        } else if (str == "p") {
+          std::cout << ss.str() << std::endl;
+        } else if (str == "s") {
+          InspectorManager::get()->check_value_skip_all_ = true;
+        }
+      }
+    }
+  }
+
+  /*!
+   * \brief infer the python type, given the c++ type
+   * \tparam ti the type info 
+   */
+  inline char infer_type(const std::type_info& ti) {
+    if (ti == typeid(float)) return 'f';
+    else if (ti == typeid(double)) return 'f';
+    else if (ti == typeid(mshadow::half::half_t) ) return 'f';
+    else if (ti == typeid(uint8_t)) return 'u';
+    else if (ti == typeid(int32_t)) return 'i';
+    else if (ti == typeid(int64_t)) return 'i';
+    else
+      return '?';
+  }
+
+  /*!
+   * \brief check if the host machine is big or small endian
+   */
+  inline char endian_test() {
+    int x = 1;
+    return (reinterpret_cast<char*>(&x)[0]) ? '<' : '>';
+  }
+
+  /*!
+   * \brief dump the value of the tensor to a file with name "[tag]_[visit count].npy" in npy format
+   * the dump file follows npy 1.0 stantand
+   * \tparam DType the data type
+   * \param tag the name given to this call
+   */
+  template<typename DType MSHADOW_DEFAULT_DTYPE>
+  inline void dump_value_helper(const std::string& tag) {
+#if MXNET_USE_CUDA
+    if (tb_.dev_mask() == gpu::kDevMask) {
+      TensorInspector(test::CAccessAsCPU(ctx_, tb_, false)(), ctx_)
+          .dump_value_helper<DType>(tag);
+      return;
+    }
+#endif  // MXNET_USE_CUDA
+    std::string dict;
+    dict += "{'descr':'";
+    dict += endian_test();
+    dict += infer_type(typeid(DType));
+    dict += std::to_string(sizeof(DType));
+    dict += "','fortran_order':False,'shape':(";
+    dict += std::to_string(tb_.shape_[0]);
+    for (int i = 1; i < tb_.ndim(); ++i) {
+      dict += ',';
+      dict += std::to_string(tb_.shape_[i]);
+    }
+    if (tb_.ndim() == 1) {
+       dict += ",";
+    }
+    dict += ")} ";
+    int padding_size = 64 - ((10 + dict.size()) % 64);
+    dict += std::string(padding_size, ' ');
+    dict[dict.size()-1] = '\n';
```

Comment:
Consider using .back() for more clarity.

Reply:
fixed